// @azure/event-processor-host/dist-esm/src/baseLease.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
/**
 * Describes the base lease.
 */
export class BaseLease {
    /**
     * @constructor
     * @param info The information required to create a base lease.
     */
    constructor(info) {
        /**
         * @property {boolean} isOwned Indicates wether the lease is owned. `true` if it is owned by
         * someone; `false` otherwise.
         */
        this.isOwned = false;
        this.partitionId = info.partitionId;
        this.owner = info.owner;
    }
    /**
     * Compares possibleOwner against this.owner
     * @param {string} possibleOwner The owner name to check.
     * @returns {boolean} boolean - true if possibleOwner is the same as this.owner, false otherwise.
     */
    isOwnedBy(possibleOwner) {
        return this.owner === possibleOwner;
    }
    /**
     * Gets the lease information.
     * @returns {CompleteLeaseInfo} CompleteLeaseInfo.
     */
    getInfo() {
        const info = {
            partitionId: this.partitionId,
            owner: this.owner
        };
        return info;
    }
}
